src/persist-client/src/operators/shard_source.rs

```diff
@@ -406,9 +441,31 @@ where
                 }
             }
         }
+
+        // We have finished outputting all leased parts the dataflow will need,
+        // but `shard_source_fetches` may still be fetching those parts. We must
+        // keep our Subscribe (and its SeqNo hold) alive until all of them have
+        // been fetched to avoid racing with GC and panicking on a missing blob.
+        // We can drop our handle when the feedback edge from `shard_source_fetches`
+        // advances to the empty frontier, indicating that all parts have been read.
+        while let Some(completed_fetch) = completed_fetches.next().await {
+            match completed_fetch {
+                Event::Data(_cap, parts) => {
+                    parts.swap(&mut parts_buffer);
+                    for part in parts_buffer.drain(..) {
+                        lease_returner.return_leased_part(lease_returner.leased_part_from_exchangeable::<G::Timestamp>(part));
+                    }
+                }
+                Event::Progress(frontier) => {
```

Comment:
Not 100% sure but it might be that we already saw the frontier on this input advance to `[]` in the loop above. We might want to check the current frontier before we go into this here loop.

Reply:
check the frontier of the current operator?

Reply:
yes, I think we might have to store the current frontier ourselves, and update it when we see a `Progress` message for the feedback in the first event loop.